riscv-debug-spec.tex

```diff
@@ -242,39 +259,80 @@ \subsubsection{Serial Ports}
-client).  For this use case the debug interface may include simple serial
-ports.
-
-A third use for the external debug interface is to use it as a general
-transport for a component to communicate with the outside world. For instance,
-it could be used to implement a serial interface that firmware could use to
-provide a simple CLI. This can use the same serial ports used for run-mode
-debugging.
+There are several use cases for dedicated debugging hardware, both
+internal to a CPU core and with an external connection.
+This specification defines techniques to support all of the use cases
+listed below. Some are
+optional to allow system designers to make cost vs capability tradeoffs.
+
+\begin{itemize}
+
+\item Debugging low-level software in the absence of an OS or other software.
+
+\item Debugging issues in the OS itself.
+
+\item Bootstrapping a system to test, configure, and program components before
+  there is any executable code path in the system.
+
+\item Accessing hardware on the system without a working CPU.
+
+\item Accesssing parts of a system which a CPU may not have software visibility
+  (e.g. cache tag arrays).
+
+\item Efficiently communicating data between a running process and external hardware,
+  for example to download code and data to write a program to memory.
+
+\item Analyzing low-level performance issues using profiling and sampling techniques.
+
+\item Providing a general transport for firmware running on a component to communicate with the
+  outside world.
+
+\end{itemize}
+
+In addition, even without a hardware debugging interface,
+architectural support in a RISC-V CPU can aid software debugging and
+performance analysis, by allowing hardware triggers and breakpoints.
+This specification aims to define common resources which can be used
+for different cases.
+
+When debugging software, this specification distinguishes between two
+forms of external debugging.
+The first is \emph{halt mode} debugging, where an external debugger halts
+some or all components of a
+platform and inspects their state while they are in stasis.
+The debugger can read and/or modify state, then
+direct the hardware
+to execute a single instruction, or continue to run freely.
+
+The second is \emph{run mode} debugging.
+In this mode a software
+debug agent runs on a component (eg.  triggered by a timer interrupt or breakpoint
+on a RISC-V core) which transfers data to or from the debugger
+without halting the component, only briefly interrupting its program flow.
+This functionality is essential if the component is controlling some real-time system (like a
+hard drive) where long timing delays could lead to physical damage.
+This requires additional software support (both on the system as well as on the debugger),
+and efficient communication channels between the component and the debugger.
+
+%MAW -- where does Quick Access fit in the above description? Is it fair to call
+% it run-mode debugging?
```

Comment:
I think of it as an optimization on regular halt-mode debugging, but you are right that it enables the things you mentioned. Without some kind of hardware quick access support, a debugger would have a major impact on the system it's polling.